pgml-sdks/python/pgml/pgml/collection.py

```diff
@@ -0,0 +1,771 @@
+                % (splitter_name, splitter_params, self.splitters_table)
+            )
+        else:
+            insert_statement = "INSERT INTO %s (name, parameters) VALUES (%s, %s)" % (
+                self.splitters_table,
+                sql.Literal(splitter_name).as_string(conn),
+                sql.Literal(json.dumps(splitter_params)).as_string(conn),
+            )
+            run_create_or_insert_statement(conn, insert_statement)
+            results = run_select_statement(conn, select_statement)
+        self.pool.putconn(conn)
+
+        return results[0]["id"]
+
+    def get_text_splitters(self) -> List[Dict[str, Any]]:
+        """
+        This function retrieves a list of dictionaries containing information about text splitters from a
+        database.
+        :return: The function `get_text_splitters` is returning a list of dictionaries, where each
+        dictionary contains the `id`, `name`, and `parameters` of a text splitter.
+        """
+        conn = self.pool.getconn()
+        select_statement = "SELECT id, name, parameters FROM %s" % (
+            self.splitters_table
+        )
+        splitters = run_select_statement(conn, select_statement)
+        self.pool.putconn(conn)
+        return splitters
+
+    def generate_chunks(self, splitter_id: int = 1) -> None:
+        """
+        This function generates chunks of text from unchunked documents using a specified text splitter.
+
+        :param splitter_id: The ID of the splitter to use for generating chunks, defaults to 1
+        :type splitter_id: int (optional)
+        """
+        conn = self.pool.getconn()
+        log.info("Using splitter id %d" % splitter_id)
+        select_statement = "SELECT name, parameters FROM %s WHERE id = %d" % (
+            self.splitters_table,
+            splitter_id,
+        )
+        results = run_select_statement(conn, select_statement)
+        splitter_name = results[0]["name"]
+        splitter_params = results[0]["parameters"]
+        if splitter_name == "RecursiveCharacterTextSplitter":
+            text_splitter = RecursiveCharacterTextSplitter(**splitter_params)
+        else:
+            raise ValueError("%s is not supported" % splitter_name)
+
+        select_statement = (
+            "SELECT id, text FROM %s WHERE id NOT IN (SELECT document FROM %s WHERE splitter = %d);"
+            % (self.documents_table, self.chunks_table, splitter_id)
+        )
+
+        results = run_select_statement(conn, select_statement)
+        for result in track(results, description="Generating chunks"):
+            log.debug("Came into chunk insertion")
+            document = result["id"]
+            text = result["text"]
+            chunks = text_splitter.create_documents([text])
+            for chunk_id, chunk in enumerate(chunks):
+                insert_statement = (
+                    "INSERT INTO %s (document,splitter,chunk_id, chunk) VALUES (%s, %s, %s, %s);"
```

Comment:
There seems to be something else going on here because it looks like we store an `id` and a `chunk_id` and the chunk text and this is the chunks table. Is the `chunk_id` actually extraneous?

Reply:
Yes, `chunk_id` is the chunk index for the given document text. For example: if the text "hello world" is split into two chunks then chunk_id 0 will map to "hello" and chunk_id 1 will map to "world". `id` is the global id of the chunk across all documents and splitters.

Reply:
Yes, this is a perfect example of needing both a local and reference to the same "object" in a single context, and being forced to disambiguate one of them with an `_id` suffix. The confusing ones in this line are `document`, and `splitter`. `chunk` and `chunk_id` are clear and easily distinguished, even though `chunk_id` does not in fact refer to a foreign key in a `chunks` table, it's clearly marked as a reference.

Reply:
Let's look at the schema in full:

```
CREATE TABLE IF NOT EXISTS %s (
    id          serial8 PRIMARY KEY,
    created_at  timestamptz NOT NULL DEFAULT now(),
    document    int8 NOT NULL REFERENCES
                     ON DELETE CASCADE
                     ON UPDATE CASCADE
                     DEFERRABLE INITIALLY DEFERRED,
    splitter    int8 NOT NULL REFERENCES %s
                     ON DELETE CASCADE
                     ON UPDATE CASCADE
                     DEFERRABLE INITIALLY DEFERRED,
    chunk_id    int8 NOT NULL,
    chunk       text NOT NULL
);
```

What is `chunk_id` actually referencing? There are some global chunks somewhere? Or is it the ordering of the chunk within the document? If it is the ordering, it should be called something else, like `chunk_index` or `position`.

Reply:
`chunk_id` is the ordering of the chunk within the document. `chunk_index` sounds good.